edi_storage/components/listener.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2021 ForgeFlow S.L. (https://www.forgeflow.com)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+import functools
+from pathlib import PurePath
+
+from odoo.addons.component.core import Component
+
+
+class EdiStorageListener(Component):
+    _name = "edi.storage.component.listener"
+    _inherit = "base.event.listener"
+
+    def _move_file(self, storage, from_dir_str, to_dir_str, filename):
+        from_dir = PurePath(from_dir_str)
+        to_dir = PurePath(to_dir_str)
+        if filename not in storage.list_files(from_dir):
+            # The file might have been moved after a previous error.
+            return False
+        self._add_after_commit_hook(
+            storage._move_files, [(from_dir / filename).as_posix()], to_dir.as_posix()
+        )
+        return True
+
+    def _add_after_commit_hook(self, move_func, sftp_filepath, sftp_destination_path):
+        """Add hook after commit to move the file when transaction is over.
+        """
+        self.env.cr.after(
+            "commit",
+            functools.partial(move_func, sftp_filepath, sftp_destination_path),
+        )
+
+    def on_edi_exchange_done(self, record):
+        storage = record.backend_id.storage_id
+        res = False
+        if record.direction == "input" and storage:
+            file = record.exchange_filename
+            pending_dir = record.backend_id.input_dir_pending
+            done_dir = record.backend_id.input_dir_done
+            error_dir = record.backend_id.input_dir_error
+            res = self._move_file(storage, pending_dir, done_dir, file)
+            if not res:
+                # If a file previously failed it should have been previously
```

Comment:
Question: if it landed in the error dir it means that corrections were needed. Hence, I would expect to be fixed and move to pending, no?

Reply:
But the file is already fetched and in the file content of exchange record. Therefore it can happen that the processing initially failed due to a misconfiguration in any place in Odoo (a tax wrongly encoded for saying something), and after fixing the configuration in Odoo [the next time that exchange record is processed](https://github.com/OCA/edi/blob/13.0/edi/models/edi_backend.py#L550) it will succeed. In that situation the file is in the error folder and need to be moved to the done folder.

Reply:
I see. Thanks for the explanation :+1: